### PR TITLE
Make use of SHA-1 and MD5 conditional, off by default

### PR DESCRIPTION
**Pull request by kontura**

### CODE REVIEW

**Comment by kontura** [comment]
Tests update: https://github.com/rpm-software-management/ci-dnf-stack/pull/987

File: CMakeLists.txt [hunk 1]
```diff
@@ -67,6 +67,12 @@ IF (ENABLE_LEGACY_WEAKDEPS)
     ADD_DEFINITIONS("-DENABLE_LEGACY_WEAKDEPS=1")
 ENDIF (ENABLE_LEGACY_WEAKDEPS)
 
+# Legacy hash functions
+OPTION (ENABLE_LEGACY_HASH "Build with SHA-1 and MD5 support" OFF)
```
Comment (dmach) [inline CMakeLists.txt:71]: I'd consider renaming `ENABLE_LEGACY_HASH` to `WITH_LEGACY_HASHES`

File: src/checksum.c [hunk 2]
```diff
@@ -52,28 +52,27 @@ cr_checksum_type(const char *name)
...
         else if (!strcmp(sha_type, "512"))
             return CR_CHECKSUM_SHA512;
+#ifdef ENABLE_LEGACY_HASH
+        else if (!strcmp(sha_type, ""))
+            return CR_CHECKSUM_SHA;
+        else if (!strcmp(sha_type, "1"))
+            return CR_CHECKSUM_SHA1;
+    } else if (!strncmp(name_lower, "md", 2)) {
+        // MD* family
+        if (name_lower[2] == '5')
```
Comment (dmach) [inline src/checksum.c:73]: Consider reordering hashes - the strongest first.
in the future you'll just move the #ifdef up to move weakest among the legacy hashes.

**Review by dmach — CHANGES_REQUESTED** [review]

File: src/python/createrepo_c/__init__.py [hunk 3]
```diff
@@ -17,9 +17,12 @@
...
 CHECKSUM_UNKNOWN    = _createrepo_c.CHECKSUM_UNKNOWN #: Checksum unknown
-MD5                 = _createrepo_c.MD5              #: MD5 checksum
-SHA                 = _createrepo_c.SHA              #: SHA1 checksum alias
-SHA1                = _createrepo_c.SHA1             #: SHA1 checksum
+if hasattr(_createrepo_c, 'MD5'):
+    MD5             = _createrepo_c.MD5              #: MD5 checksum
+if hasattr(_createrepo_c, 'SHA'):
+    SHA             = _createrepo_c.SHA              #: SHA1 checksum alias
+if hasattr(_createrepo_c, 'SHA1'):
+    SHA1            = _createrepo_c.SHA1             #: SHA1 checksum
```
Comment (dmach) [inline src/python/createrepo_c/__init__.py:25]: Couldn't this be made more future-proof?
```
for hash_name in ('MD5', 'SHA', 'SHA1', 'SHA224', 'SHA256', 'SHA512'):
    hash = getattr(_createrepo_c, hash_name, None)
    if hash:
        globals()[hash_name] = hash
```

Comment (kontura) [reply src/python/createrepo_c/__init__.py]: That is neat, thanks.

**Comment by kontura** [comment]
I have also added numbering to the `cr_ChecksumType` enum to preserve the values.

**Comment by kontura** [comment]
I have added one more commit, fixing a possible memory leak.

**Comment by kontura** [comment]
I have removed the last added commit because in the meantime it was fixed by https://github.com/rpm-software-management/createrepo_c/pull/267.

File: CMakeLists.txt [hunk 4]
```diff
@@ -67,6 +67,12 @@ IF (ENABLE_LEGACY_WEAKDEPS)
     ADD_DEFINITIONS("-DENABLE_LEGACY_WEAKDEPS=1")
 ENDIF (ENABLE_LEGACY_WEAKDEPS)
 
+# Legacy hash functions
+OPTION (WITH_LEGACY_HASHES "Build with SHA-1 and MD5 support" OFF)
+IF (WITH_LEGACY_HASHES)
+    ADD_DEFINITIONS("-DWITH_LEGACY_HASHES=1")
+ENDIF (WITH_LEGACY_HASHES)
```
Comment (dralley) [inline CMakeLists.txt:74]: Could you add a line to `setup.py` [here](https://github.com/rpm-software-management/createrepo_c/blob/master/setup.py#L45) which turns it ON for the Python build?

Comment (kontura) [reply CMakeLists.txt:74]: Yes, added.

**Comment by jrohel** [comment]
The condition is resolved too soon. During compilation, not at runtime. But we discussed the implications, so I'm going to merge it.

**Comment by praiskup** [comment]
FTR, EL5 systems seem to only support md5 checksums in repo metadata, so this will
disallow people on new systems to generate metadata for ancient systems.  Detected by Copr
testsuite: https://pagure.io/copr/copr/issue/1971

**Comment by dralley** [comment]
re: ^, Pulp / Satellite enables md5 in our build for this reason.

Temporarily, we also have a patch that allows it to work when running on FIPS-enabled hosts.  Hoping to get rid of it soon, it's needed mostly for migrating those repos from Pulp 2.

```
diff --git a/src/checksum.c b/src/checksum.c
index 485461d..6943886 100644
--- a/src/checksum.c
+++ b/src/checksum.c
@@ -145,6 +145,13 @@ cr_checksum_file(const char *filename,
     }
 
     ctx = EVP_MD_CTX_create();
+    if (!ctx) {
+        g_set_error(err, ERR_DOMAIN, CRE_OPENSSL,
+                    "EVP_MD_CTX_create() failed");
+        return NULL;
+    }
+    EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_NON_FIPS_ALLOW);
+
     rc = EVP_DigestInit_ex(ctx, ctx_type, NULL);
     if (!rc) {
         g_set_error(err, ERR_DOMAIN, CRE_OPENSSL,
@@ -229,6 +236,7 @@ cr_checksum_new(cr_ChecksumType type, GError **err)
                     "EVP_MD_CTX_create() failed");
         return NULL;
     }
+    EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_NON_FIPS_ALLOW);
 
     if (!EVP_DigestInit_ex(ctx, ctx_type, NULL)) {
         g_set_error(err, ERR_DOMAIN, CRE_OPENSSL,

```

**Pull request merged**